src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -6,193 +6,592 @@
+            else if (encoder.IsInvariantUtf16)
+                _encoderState = JsonEncoderState.UseFastUtf16;
+            else
+                _encoderState = JsonEncoderState.UseFullEncoder;
+        }
 
+        /// <summary>
+        /// Write the starting tag of an object. This is used for adding an object to an 
+        /// array of other items. If this is used while inside a nested object, the property
+        /// name will be missing and result in invalid JSON.
+        /// </summary>
         public void WriteObjectStart()
         {
-            _wroteListItem = false;
-            _formatter.Append('{');
-            WriteNewline();
+            WriteItemSeperator();
+            WriteSpacing(false);
+            WriteChar(JsonConstants.OpenBrace);
+
+            _firstItem = true;
             _indent++;
         }
 
+        /// <summary>
+        /// Write the starting tag of an object. This is used for adding an object to a
+        /// nested object. If this is used while inside a nested array, the property
+        /// name will be written and result in invalid JSON.
+        /// </summary>
+        /// <param name="name">The name of the property (i.e. key) within the containing object.</param>
+        public void WriteObjectStart(string name)
+        {
+            WriteStartAttribute(name);
+            WriteChar(JsonConstants.OpenBrace);
+
+            _firstItem = true;
+            _indent++;
+        }
+
+        /// <summary>
+        /// Writes the end tag for an object.
+        /// </summary>
         public void WriteObjectEnd()
         {
+            _firstItem = false;
             _indent--;
-            WriteNewline();
-            WriteIndent();
-            _formatter.Append('}');
-            _wroteListItem = true;
+            WriteSpacing();
+            WriteChar(JsonConstants.CloseBrace);
         }
 
+        /// <summary>
+        /// Write the starting tag of an array. This is used for adding an array to a nested
+        /// array of other items. If this is used while inside a nested object, the property
+        /// name will be missing and result in invalid JSON.
+        /// </summary>
         public void WriteArrayStart()
         {
-            _wroteListItem = false;
-            _formatter.Append('[');
-            WriteNewline();
+            WriteItemSeperator();
+            WriteSpacing();
+            WriteChar(JsonConstants.OpenBracket);
+
+            _firstItem = true;
             _indent++;
         }
 
-        public void WriteArrayEnd()
+        /// <summary>
+        /// Write the starting tag of an array. This is used for adding an array to a
+        /// nested object. If this is used while inside a nested array, the property
+        /// name will be written and result in invalid JSON.
+        /// </summary>
+        /// <param name="name">The name of the property (i.e. key) within the containing object.</param>
+        public void WriteArrayStart(string name)
```

Comment:
It seems cleaner. It seemed odd (and definitely not as efficient) to have to call WritePropertyName + WriteValue / WriteObjectStart / WriteArrayStart for every property. Combining them into one call was a performance bump and made the resulting code easier to read / parse in my head.

If we have a burning desire to keep the pattern from JSON.NET, then I can be convinced to change it. However, considering this is supposed to be an efficient / low-level layer, I liked this pattern.